File: cloudera_training/devsh/exercises/spark-streaming-multi/solution-python/StreamingLogsMB.py
import sys
from pyspark import SparkContext
from pyspark.streaming import StreamingContext

# Given an array of new counts, add up the counts 
# and then add the sum to the old counts and return the new total
def updateCount(newCounts, state): 
    if state == None: return sum(newCounts)
    else: return state + sum(newCounts)

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print >> sys.stderr, "Usage: StreamingLogsMB.py <hostname> <port>"
        sys.exit(-1)
    
    # get hostname and port of data source from application arguments
    hostname = sys.argv[1]
    port = int(sys.argv[2])
     
    # Create a new SparkContext
    sc = SparkContext()

    # Set log level to ERROR to avoid distracting extra output
    sc.setLogLevel("ERROR")

    # Create and configure a new Streaming Context 
    # with a 1 second batch duration
    ssc = StreamingContext(sc,1)

    # Create a DStream of log data from the server and port specified    
    logs = ssc.socketTextStream(hostname,port)

    # Every two seconds, display the total number of requests over the 
    # last 5 seconds
    logs.countByWindow(5,2).pprint()
    
    # Bonus: Display the top 5 users every second

    # Enable checkpointing (required for state operations)
    ssc.checkpoint("logcheckpt")
    
    # Count requests by user ID for every batch
    userreqs = logs \
        .map(lambda line: (line.split(" ")[2],1)) \
        .reduceByKey(lambda v1,v2: v1+v2)

    # Update total user requests
    totalUserreqs = userreqs \
        .updateStateByKey(lambda newCounts, state: updateCount(newCounts, state))

    # Sort each state RDD by hit count in descending order    
    topUserreqs=totalUserreqs \
       .map(lambda (k,v):(v,k)) \
       .transform(lambda rdd: rdd.sortByKey(False)) \
       .map(lambda (k,v):(v,k)) 

    topUserreqs.pprint()
    
   
    ssc.start()
    ssc.awaitTermination()

File: cloudera_training/devsh/exercises/spark-streaming-multi/streaminglogsMB_project/src/main/scala/solution/StreamingLogsMB.scala
package solution

import org.apache.spark.SparkContext
import org.apache.spark.streaming.StreamingContext
import org.apache.spark.streaming.Seconds

object StreamingLogsMB {
  
  // Given an array of new counts, add up the counts 
  // and then add them to the old counts and return the new total
  def updateCount = (newCounts: Seq[Int], state: Option[Int]) => {
     val newCount = newCounts.foldLeft(0)(_ + _)
     val previousCount = state.getOrElse(0)
     Some(newCount + previousCount) 
  }
  
  
  def main(args: Array[String]) {
    if (args.length < 2) {
      System.err.println("Usage: solution.StreamingLogsMB <hostname> <port>")
      System.exit(1)
    } 
 
    // get hostname and port of data source from application arguments
    val hostname = args(0)
    val port = args(1).toInt

    // Create a Spark Context
    val sc = new SparkContext()

    // Set log level to ERROR to avoid distracting extra output
    sc.setLogLevel("ERROR")

    // Configure the Streaming Context with a 1 second batch duration
    val ssc = new StreamingContext(sc,Seconds(1))

    // Create a DStream of log data from the server and port specified   
    val logs = ssc.socketTextStream(hostname,port)

    // Every two seconds, display the total number of requests over the 
    // last 5 seconds
    logs.countByWindow(Seconds(5),Seconds(2)).print()

    // Bonus: Display the top 5 users every second

    // Enable checkpointing (required for all window and state operations)
    ssc.checkpoint("logcheckpt")
    
    // Count requests by user ID for every batch
    val userreqs = logs.
       map(line => (line.split(" ")(2),1)).
       reduceByKey((x,y) => x+y)

    // Update total user requests
    val totalUserreqs = userreqs.updateStateByKey(updateCount)

    // Sort each state RDD by hit count in descending order    
    val topUserreqs=totalUserreqs.
       map(pair => pair.swap).
       transform(rdd => rdd.sortByKey(false)).
       map(pair => pair.swap)

    topUserreqs.print()
    


    ssc.start()
    ssc.awaitTermination()
  }

}